includes/gateways/paypal/admin/connect.php

```diff
@@ -0,0 +1,478 @@
+			'code'          => $_POST['auth_code'],
+			'code_verifier' => $partner_details->nonce
+		)
+	) );
+
+	if ( is_wp_error( $response ) ) {
+		wp_send_json_error( $response->get_error_message() );
+	}
+
+	$code = wp_remote_retrieve_response_code( $response );
+	$body = json_decode( wp_remote_retrieve_body( $response ) );
+
+	if ( empty( $body->access_token ) ) {
+		wp_send_json_error( sprintf(
+		/* Translators: %d - HTTP response code */
+			__( 'Unexpected response from PayPal while generating token. Response code: %d. Please try again.', 'easy-digital-downloads' ),
+			$code
+		) );
+	}
+
+	/*
+	 * Now we can use this access token to fetch the seller's credentials for all future
+	 * API requests.
+	 */
+	$response = wp_remote_get( $api_url . 'v1/customer/partners/' . urlencode( \EDD\PayPal\get_partner_merchant_id( $mode ) ) . '/merchant-integrations/credentials/', array(
+		'headers' => array(
+			'Authorization' => sprintf( 'Bearer %s', $body->access_token ),
+			'Content-Type'  => 'application/json',
+			'timeout'       => 15
+		)
+	) );
+
+	if ( is_wp_error( $response ) ) {
+		wp_send_json_error( $response->get_error_message() );
+	}
+
+	$code = wp_remote_retrieve_response_code( $response );
+	$body = json_decode( wp_remote_retrieve_body( $response ) );
+
+	if ( empty( $body->client_id ) || empty( $body->client_secret ) ) {
+		wp_send_json_error( sprintf(
+		/* Translators: %d - HTTP response code */
+			__( 'Unexpected response from PayPal. Response code: %d. Please try again.', 'easy-digital-downloads' ),
+			$code
+		) );
+	}
+
+	edd_update_option( 'paypal_' . $mode . '_client_id', sanitize_text_field( $body->client_id ) );
+	edd_update_option( 'paypal_' . $mode . '_client_secret', sanitize_text_field( $body->client_secret ) );
+
+	// We don't need the nonce anymore so we can delete that.
+	delete_option( 'edd_paypal_commerce_connect_details' );
+
+	$message = esc_html__( 'Successfully connected.', 'easy-digital-downloads' );
+
+	try {
+		PayPal\Webhooks\create_webhook( $mode );
+	} catch ( \Exception $e ) {
+		$message = esc_html__( 'Your account has been successfully connected, but an error occurred while creating a webhook.', 'easy-digital-downloads' );
+	}
+
+	/**
+	 * Triggers when an account is successfully connected to PayPal.
+	 *
+	 * @param string $mode The mode that the account was connected in. Either `sandbox` or `live`.
+	 *
+	 * @since 2.11
+	 */
+	do_action( 'edd_paypal_commerce_connected', $mode );
+
+	wp_send_json_success( $message );
+}
+
+add_action( 'wp_ajax_edd_paypal_commerce_get_access_token', __NAMESPACE__ . '\get_and_save_credentials' );
+
+/**
+ * Verifies the connected account.
+ *
+ * @since 2.11
+ * @return void
+ */
+function get_account_info() {
+	check_ajax_referer( 'edd_paypal_account_information' );
+
+	if ( ! current_user_can( 'manage_shop_settings' ) ) {
+		wp_send_json_error( wpautop( __( 'You do not have permission to perform this action.', 'easy-digital-downloads' ) ) );
```

Comment:
The `NaN` is coming from here, or a place like it--because instead of sending an array of data back to the `eddPayPalGetAccountStatus` function in the JS, it's just sending back the JSON object with `success: false` and then the `data` is this message as a string.

I suspect other uses of `wp_send_json_error` may have the same issue--my guess would be that the JS function(s) need to be updated to have both a `success` and `fail` function, maybe? Or the checks and output need to be adjusted in the JS to account for `data` as a string with an error message.

I was able to trigger this by just removing the `!` in line 251.

Reply:
Using [`wp.ajax.send()`](https://github.com/WordPress/WordPress/blob/master/wp-includes/js/wp-util.js#L82) allows a `success` and `error` callback to be supplied via the `options` object and will be called based on `wp_send_json_error()` or `wp_send_json_success()`.

Reply:
Updated!